### DISCUSSION TITLE
Non-global npm install

### DISCUSSION THREAD

**Discussion by xverges** [TiddlyWiki/TiddlyWiki5#5420] [Ideas]
👋 Back to TiddlyWiki after several years, trying to learn again.

Once upon a time, before browsers made it hard to write to the local filesystem even for `file:` urls, the TiddlyWiki on-boarding experience was really simple and lovely. Unfortunately, now it needs to be necessarily harder. I am now finding my way in node.js-based on-boarding.

As a developer, I always try to avoid global npm installs. @Arlen22's TiddlyServer (that I have come to understand should be my tool of choice to run several TiddlyWikis with on single server and port) [does document a non-global install](https://arlen22.github.io/tiddlyserver/docs/gettingstarted#portable-install), but does make access to the `tiddlywiki` command cumbersome, and even more to execute something equivalent to `tiddlywiki --editions`.

Bugged by this, today I learned about the [`initializer` optional param to `npm init`](https://docs.npmjs.com/cli/v6/commands/npm-init#description). `npm init tiddlywiki [extra params]` would execute `npx create-tiddlywiki [extra params]`, where the `main` of the `create-tiddlywiki` npm package would be creating a specially crafted `package.json`, where  appropriate  `npm run` scripts could be added (and [bash autocompletion for them would be provided by `npm`](https://github.com/npm/npm/issues/8030))

Would it make sense to adopt this pattern, so that users are not _forced_ (*) to have a single version of the `tiddlywiki` command? 

*: I understand that they are not forced, and that knowledgeable users can use a local install

**Comment by inmysocks** [comment]
I am not sure what is easier about this than just installing tiddlywiki without the -g flag. The tiddlywiki commands require a path to the wiki folder anyway.

For my computers I clone the tiddlywiki5 repo and then have the wikis in a folder next to it, like in the https://github.com/inmysocks/PluginDevelopment repo.

**Comment by Arlen22** [comment]
Would it work to do a local install in exactly the same way as outlined for TiddlyServer? You would then put your custom commands in the scripts object of your package.json and if you pass additional args they should be passed in. 

```json
"scripts": {
  "tiddlyserver": "tiddlyserver",
  "tiddlywiki": "tiddlywiki",
  "tiddlywiki": "tiddlywiki-production-server"
}
```

I recently discovered that args are passed in, so `npm run tiddlywiki ./data --listen` should work with this.

**Oops:** It has to be `npm run tiddlywiki -- ./data --listen`. See my comment below.

`tiddlywiki-production-server` doesn't include the editions or bin folders, and compacts core and plugins into one file each to improve startup time. If you need editions or bin you can just `npm install tiddlywiki` separately and include it in the scripts object as shown.

**Reply by Arlen22** [reply]
I just discovered a serious drawback here. You can't pass flags (starting with dashes) without passing the `--` arg first. This would definitely catch people off guard. `npx` does not have this problem. I think I would probably recommend just making sure npx is installed (usually is alongside npm) or shimming it in your portable install with a cmd or bash file, if you really need it. Otherwise you can just include the path in `node_modules/.bin`.

**Comment by Arlen22** [comment]
If you don't want to depend on having npm installed globally, you could just write a script to make the proper calls to the different items and pass in the variables. Normally on windows npm includes such a script in `node_modules/.bin` named according to the executable name of any executable dependancies. On linux it just symlinks the file if possible.

**Comment by xverges** [comment]
I have created a https://github.com/xverges/create-tiddlywiki to showcase what kind of on-boarding experience we could be having if we supported `npm init tiddlywiki`. Of course, we could be asking the user more questions (e.g. find out if the user would like to use TiddlyDesktop, or TiddlyServer, or..) and give appropiate instructions or create the appropiate shell scripts in the installation folder to help the user launch those tools and the `tiddlywiki` command.

This silent 34 seconds video shows what would be the effect of running `npm init tiddlywiki`

https://user-images.githubusercontent.com/460418/105973416-ffe7c780-608c-11eb-964b-0f705b71e0a4.mp4

**Comment by xverges** [comment]
I have updated https://github.com/xverges/create-tiddlywiki and now
* it asks you if you want to create more than a TiddlyWiki
* it asks you if you want to run your TiddlyWikis with `tiddlywiki --listen` or `tiddlyserver`

I think that it provides a quite nice on-boarding experience :-)

![image](https://user-images.githubusercontent.com/460418/106200219-f0fd3400-61b6-11eb-9db0-3282ded0fb98.png)

**Comment by Jermolene** [comment]
Thanks @xverges @inmysocks @Arlen22. Our integration with NPM is pretty basic at the moment, and I'm very interested in understanding the benefits of more extensive integration. The ability to run code without installing it, and the Bash integration are both attractive features.

Perhaps the way to proceed might be a PR that includes the associated documentation updates: the big win here would indeed seem to be the possibility of radically improving the onboarding experience.

I haven't read through all of the code, but it seems like there's more than just glue; there's the code of a setup wizard, for example. Can those things be structured as ordinary TW5 commands so that they are available for people not using npm?

**Reply by xverges** [reply]
Very glad that you are jumping in, @Jermolene !

>  haven't read through all of the code, but it seems like there's more than just glue; there's the code of a setup wizard, for example. Can those things be structured as ordinary TW5 commands so that they are available for people not using npm?

The code is intended to be executed by calling `npm init tiddlywiki`, but, since it has not been published in the `npm` registry, it has to be executed now with `npx git+https://github.com/xverges/create-tiddlywiki.git`. You can give it a try in an empty folder.

Currently, it
* asks the user for folder paths and TW editions where they want to have a new TiddlyWiki initialized, and runs `tiddlywiki --init` on each of them
* asks the user if he is interested in using `tiddlywiki --listen` and/or `tiddlyserver`
* optionally, creates a configuration file for `tiddlyserver`
* creates a `package.json` file, ready for `npm install`, and later for `npm start`, that will launch `tiddlywiki --listen`
 or `tiddlyserver`

I lack perspective, but all these tasks look to me like quite `node/npm` related. Which of them would you like to see implemented as TW5 commands?

> Perhaps the way to proceed might be a PR that includes the associated documentation updates

👍 I'll see what can I do about it next week

**Reply by Jermolene** [reply]
Hi @xverges perhaps we could put most of that functionality under a new `--wizard` command, and then the npm command would be a shortcut to it?

**Reply by xverges** [reply]
@Jermolene I don't think I like the `--wizard` idea. If I am not mistaken, all tiddlywiki commands require some knowledge of the internal tiddlywiki structure/architecture. I also see that tiddlywiki does not have any external dependency in its `package.json`. In my opinion, a `--wizard` command:
* would break cohesiveness in the tiddlywiki command set
* would require to add the first dependencies to tiddlywiki's `package.json` (or be implemented without dependencies and with less features/more complexity that it needed to be).

I think this wizard is better implemented as a wrapper around `tiddlywiki.js` than as part of tiddlywiki itself. This wrapper project can potentially grow to help users accomplish other cli-based tiddlywiki tasks, while keeping the base cli focused.

**Reply by Jermolene** [reply]
Hi @xverges 

> If I am not mistaken, all tiddlywiki commands require some knowledge of the internal tiddlywiki structure/architecture.

I'm not sure what you mean. A command like "---version" doesn't require knowledge of TW structure or architecture.

> I also see that tiddlywiki does not have any external dependency in its `package.json`.

That's intentional so that people can use TW5 without having to install npm (which has many dependencies).

Perhaps that's the fundamental issue here: I'm trying to avoid making TW5 depend on npm. That means that all the functionality that we expose through npm must also be exposed in some other way. TiddlyWiki commands give us a consistent way to invoke functionality on the command line; introducing another mechanism would only complicate things.

> In my opinion, a `--wizard` command:
> 
> * would break cohesiveness in the tiddlywiki command set

I'm not sure why. The `--wizard` command would walk users through setting themselves up to use TiddlyWiki. The end result would be creating a folder and some files, just like the existing `--init` command.

> * would require to add the first dependencies to tiddlywiki's `package.json` (or be implemented without dependencies and with less features/more complexity that it needed to be).

Why would the `--wizard` command require dependencies?

> I think this wizard is better implemented as a wrapper around `tiddlywiki.js` than as part of tiddlywiki itself. This wrapper project can potentially grow to help users accomplish other cli-based tiddlywiki tasks, while keeping the base cli focused.

That really makes it sounds like a separate, external thing to TiddlyWiki.

**Reply by xverges** [reply]
> Perhaps that's the fundamental issue here: I'm trying to avoid making TW5 depend on npm

TW5 is not dependant on node or npm, and I agree it is a good decission. However, users are asked to install command line `tiddlywiki` using `npm` (again, a good decision to me).  

> Why would the --wizard command require dependencies?

Not necessarily, but without dependencies or reinventing the wheel, you would not achieve the same user experience that is provided by existing packages like https://github.com/terkelg/prompts or https://github.com/lukeed/kleur.

> That really makes it sounds like a separate, external thing to TiddlyWiki.

Agree. But it has community implications, in the form of `npm` registry namespace. Would it be ok if I registered the npm package `create-tiddlywizard`, that would enable end users to execute `npm init tiddlywizard`? That still leaves `create-tiddlywiki` and `npm init tiddlywiki` available for an official implementation.
